// victory-bar/lib/v37/victory-bar-v1.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
exports.withContainer = withContainer;

var React = _interopRequireWildcard(require("react"));

var _victoryCore = require("victory-core");

var _v = require("victory-core/es/v37");

var _bar = _interopRequireDefault(require("../bar"));

var _helperMethods = require("../helper-methods");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

// This is a demo component that uses VictoryProvider to access calculated props and state.
// This component does not include events, animations, or styling.
// To test out, swap out the VictoryBar export in ./index.js and run `pnpm run storybook:server`.
var defaultProps = {
  data: [{
    x: 1,
    y: 1
  }, {
    x: 2,
    y: 2
  }, {
    x: 3,
    y: 3
  }, {
    x: 4,
    y: 4
  }],
  height: 300,
  includeZero: true,
  padding: 50,
  sortOrder: "ascending",
  width: 450
};

// TODO: This would be a shared helper that allows us to access context values in the base component
function withContainer(WrappedComponent) {
  var initialProviderProps = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return function (props) {
    var providerProps = _objectSpread(_objectSpread({}, initialProviderProps), props);

    var _props$standalone = props.standalone,
        standalone = _props$standalone === void 0 ? true : _props$standalone,
        _props$containerCompo = props.containerComponent,
        containerComponent = _props$containerCompo === void 0 ? /*#__PURE__*/React.createElement(_victoryCore.VictoryContainer, null) : _props$containerCompo;

    if (standalone) {
      return /*#__PURE__*/React.createElement(_v.VictoryProvider, providerProps, /*#__PURE__*/React.cloneElement(containerComponent, providerProps, /*#__PURE__*/React.createElement(WrappedComponent, props)));
    }

    return /*#__PURE__*/React.createElement(_v.VictoryProvider, providerProps, /*#__PURE__*/React.createElement(WrappedComponent, props));
  };
}

function VictoryBar(_ref) {
  var _ref$dataComponent = _ref.dataComponent,
      dataComponent = _ref$dataComponent === void 0 ? /*#__PURE__*/React.createElement(_bar.default, null) : _ref$dataComponent,
      _ref$groupComponent = _ref.groupComponent,
      groupComponent = _ref$groupComponent === void 0 ? /*#__PURE__*/React.createElement("g", {
    role: "presentation"
  }) : _ref$groupComponent,
      alignment = _ref.alignment,
      barRatio = _ref.barRatio,
      barWidth = _ref.barWidth,
      cornerRadius = _ref.cornerRadius,
      _ref$horizontal = _ref.horizontal,
      horizontal = _ref$horizontal === void 0 ? false : _ref$horizontal;
  var scale = (0, _v.useScale)();
  var data = (0, _v.useData)();
  var domain = (0, _v.useDomain)();
  var children = data.map(function (datum, i) {
    var _getBarPosition = (0, _helperMethods.getBarPosition)({
      domain: domain,
      scale: scale,
      horizontal: horizontal
    }, datum),
        x = _getBarPosition.x,
        y = _getBarPosition.y,
        x0 = _getBarPosition.x0,
        y0 = _getBarPosition.y0;

    var dataProps = {
      index: i,
      key: "bar-".concat(i),
      alignment: alignment,
      barRatio: barRatio,
      barWidth: barWidth,
      cornerRadius: cornerRadius,
      scale: scale,
      data: data,
      x: x,
      y: y,
      x0: x0,
      y0: y0,
      datum: datum
    };
    return /*#__PURE__*/React.cloneElement(dataComponent, dataProps);
  });
  return /*#__PURE__*/React.cloneElement(groupComponent, {}, children);
}

var _default = withContainer(VictoryBar, defaultProps);

exports.default = _default;
