// victory-core/lib/victory-util/common-props.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CommonProps = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var CustomPropTypes = _interopRequireWildcard(require("./prop-types"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var dataProps = {
  // @ts-expect-error TODO: synchronize the type with this PropTypes
  categories: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.string), _propTypes.default.shape({
    x: _propTypes.default.arrayOf(_propTypes.default.string),
    y: _propTypes.default.arrayOf(_propTypes.default.string)
  })]),
  // @ts-expect-error TODO: synchronize the type with this PropTypes
  data: _propTypes.default.oneOfType([_propTypes.default.array, _propTypes.default.object]),
  dataComponent: _propTypes.default.element,
  disableInlineStyles: _propTypes.default.bool,
  labelComponent: _propTypes.default.element,
  labels: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.array]),
  samples: CustomPropTypes.nonNegative,
  sortKey: _propTypes.default.oneOfType([_propTypes.default.func, CustomPropTypes.allOfType([CustomPropTypes.integer, CustomPropTypes.nonNegative]), _propTypes.default.string, _propTypes.default.arrayOf(_propTypes.default.string.isRequired)]),
  sortOrder: _propTypes.default.oneOf(["ascending", "descending"]),
  style: _propTypes.default.shape({
    parent: _propTypes.default.object,
    data: _propTypes.default.object,
    labels: _propTypes.default.object
  }),
  x: _propTypes.default.oneOfType([_propTypes.default.func, CustomPropTypes.allOfType([CustomPropTypes.integer, CustomPropTypes.nonNegative]), _propTypes.default.string, _propTypes.default.arrayOf(_propTypes.default.string.isRequired)]),
  y: _propTypes.default.oneOfType([_propTypes.default.func, CustomPropTypes.allOfType([CustomPropTypes.integer, CustomPropTypes.nonNegative]), _propTypes.default.string, _propTypes.default.arrayOf(_propTypes.default.string.isRequired)]),
  y0: _propTypes.default.oneOfType([_propTypes.default.func, CustomPropTypes.allOfType([CustomPropTypes.integer, CustomPropTypes.nonNegative]), _propTypes.default.string, _propTypes.default.arrayOf(_propTypes.default.string.isRequired)])
};
var baseProps = {
  animate: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.object]),
  containerComponent: _propTypes.default.element,
  domain: _propTypes.default.oneOfType([CustomPropTypes.domain, _propTypes.default.shape({
    x: CustomPropTypes.domain,
    y: CustomPropTypes.domain
  })]),
  // @ts-expect-error TODO: synchronize the type with this PropTypes
  maxDomain: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.instanceOf(Date), _propTypes.default.shape({
    x: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.instanceOf(Date)]),
    y: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.instanceOf(Date)])
  })]),
  // @ts-expect-error TODO: synchronize the type with this PropTypes
  minDomain: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.instanceOf(Date), _propTypes.default.shape({
    x: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.instanceOf(Date)]),
    y: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.instanceOf(Date)])
  })]),
  // @ts-expect-error TODO: synchronize the type with this PropTypes
  domainPadding: _propTypes.default.oneOfType([_propTypes.default.shape({
    x: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.arrayOf(_propTypes.default.number)]),
    y: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.arrayOf(_propTypes.default.number)])
  }), _propTypes.default.number, _propTypes.default.arrayOf(_propTypes.default.number)]),
  eventKey: _propTypes.default.oneOfType([_propTypes.default.func, CustomPropTypes.allOfType([CustomPropTypes.integer, CustomPropTypes.nonNegative]), _propTypes.default.string]),
  events: _propTypes.default.arrayOf(_propTypes.default.shape({
    target: _propTypes.default.oneOf(["data", "labels", "parent"]),
    eventKey: _propTypes.default.oneOfType([_propTypes.default.array, CustomPropTypes.allOfType([CustomPropTypes.integer, CustomPropTypes.nonNegative]), _propTypes.default.string]),
    eventHandlers: _propTypes.default.object
  })),
  // @ts-expect-error TODO: synchronize the type with this PropTypes
  externalEventMutations: _propTypes.default.arrayOf(_propTypes.default.shape({
    callback: _propTypes.default.func,
    childName: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.array]),
    eventKey: _propTypes.default.oneOfType([_propTypes.default.array, CustomPropTypes.allOfType([CustomPropTypes.integer, CustomPropTypes.nonNegative]), _propTypes.default.string]),
    mutation: _propTypes.default.func,
    target: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.array])
  })),
  groupComponent: _propTypes.default.element,
  height: CustomPropTypes.nonNegative,
  name: _propTypes.default.string,
  // @ts-expect-error TODO: synchronize the type with this PropTypes
  origin: _propTypes.default.shape({
    x: _propTypes.default.number,
    y: _propTypes.default.number
  }),
  // @ts-expect-error TODO: synchronize the type with this PropTypes
  padding: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.shape({
    top: _propTypes.default.number,
    bottom: _propTypes.default.number,
    left: _propTypes.default.number,
    right: _propTypes.default.number
  })]),
  polar: _propTypes.default.bool,
  range: _propTypes.default.oneOfType([CustomPropTypes.domain, _propTypes.default.shape({
    x: CustomPropTypes.domain.isRequired,
    y: CustomPropTypes.domain.isRequired
  })]),
  scale: _propTypes.default.oneOfType([CustomPropTypes.scale, _propTypes.default.shape({
    x: CustomPropTypes.scale.isRequired,
    y: CustomPropTypes.scale.isRequired
  })]),
  // @ts-expect-error TODO: synchronize the type with this PropTypes
  sharedEvents: _propTypes.default.shape({
    events: _propTypes.default.array,
    getEventState: _propTypes.default.func
  }),
  // @ts-expect-error TODO: synchronize the type with this PropTypes
  singleQuadrantDomainPadding: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.shape({
    x: _propTypes.default.oneOfType([_propTypes.default.bool]),
    y: _propTypes.default.oneOfType([_propTypes.default.bool])
  })]),
  standalone: _propTypes.default.bool,
  theme: _propTypes.default.object,
  width: CustomPropTypes.nonNegative
};
var primitiveProps = {
  active: _propTypes.default.bool,
  ariaLabel: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.func]),
  className: _propTypes.default.string,
  clipPath: _propTypes.default.string,
  data: _propTypes.default.oneOfType([_propTypes.default.array, _propTypes.default.object]),
  desc: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.func]),
  disableInlineStyles: _propTypes.default.bool,
  events: _propTypes.default.object,
  id: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string, _propTypes.default.func]),
  index: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string]),
  // @ts-expect-error TODO: synchronize the type with this PropTypes
  origin: _propTypes.default.shape({
    x: _propTypes.default.number,
    y: _propTypes.default.number
  }),
  polar: _propTypes.default.bool,
  role: _propTypes.default.string,
  scale: _propTypes.default.oneOfType([CustomPropTypes.scale, _propTypes.default.shape({
    x: CustomPropTypes.scale,
    y: CustomPropTypes.scale
  })]),
  shapeRendering: _propTypes.default.string,
  style: _propTypes.default.object,
  tabIndex: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.func]),
  transform: _propTypes.default.string
};
var CommonProps = {
  dataProps: dataProps,
  baseProps: baseProps,
  primitiveProps: primitiveProps
};
exports.CommonProps = CommonProps;
