// victory-core/lib/victory-util/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  Axis: true,
  Collection: true,
  Data: true,
  DefaultTransitions: true,
  Domain: true,
  Events: true,
  Helpers: true,
  Hooks: true,
  Immutable: true,
  LabelHelpers: true,
  LineHelpers: true,
  Log: true,
  PointPathHelpers: true,
  PropTypes: true,
  Scale: true,
  Selection: true,
  Style: true,
  TextSize: true,
  Timer: true,
  TimerContext: true,
  Transitions: true,
  UserProps: true,
  Wrapper: true
};
exports.TextSize = exports.Style = exports.Selection = exports.Scale = exports.PropTypes = exports.PointPathHelpers = exports.Log = exports.LineHelpers = exports.LabelHelpers = exports.Immutable = exports.Hooks = exports.Helpers = exports.Events = exports.Domain = exports.DefaultTransitions = exports.Data = exports.Collection = exports.Axis = void 0;
Object.defineProperty(exports, "Timer", {
  enumerable: true,
  get: function () {
    return _timer.default;
  }
});
Object.defineProperty(exports, "TimerContext", {
  enumerable: true,
  get: function () {
    return _timerContext.default;
  }
});
exports.Wrapper = exports.UserProps = exports.Transitions = void 0;

var _addEvents = require("./add-events");

Object.keys(_addEvents).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _addEvents[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _addEvents[key];
    }
  });
});

var _Axis = _interopRequireWildcard(require("./axis"));

exports.Axis = _Axis;

var _Collection = _interopRequireWildcard(require("./collection"));

exports.Collection = _Collection;

var _commonProps = require("./common-props");

Object.keys(_commonProps).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _commonProps[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _commonProps[key];
    }
  });
});

var _Data = _interopRequireWildcard(require("./data"));

exports.Data = _Data;

var _DefaultTransitions = _interopRequireWildcard(require("./default-transitions"));

exports.DefaultTransitions = _DefaultTransitions;

var _Domain = _interopRequireWildcard(require("./domain"));

exports.Domain = _Domain;

var _Events = _interopRequireWildcard(require("./events"));

exports.Events = _Events;

var _Helpers = _interopRequireWildcard(require("./helpers"));

exports.Helpers = _Helpers;

var _Hooks = _interopRequireWildcard(require("./hooks"));

exports.Hooks = _Hooks;

var _Immutable = _interopRequireWildcard(require("./immutable"));

exports.Immutable = _Immutable;

var _LabelHelpers = _interopRequireWildcard(require("./label-helpers"));

exports.LabelHelpers = _LabelHelpers;

var _LineHelpers = _interopRequireWildcard(require("./line-helpers"));

exports.LineHelpers = _LineHelpers;

var _Log = _interopRequireWildcard(require("./log"));

exports.Log = _Log;

var _PointPathHelpers = _interopRequireWildcard(require("./point-path-helpers"));

exports.PointPathHelpers = _PointPathHelpers;

var _PropTypes = _interopRequireWildcard(require("./prop-types"));

exports.PropTypes = _PropTypes;

var _Scale = _interopRequireWildcard(require("./scale"));

exports.Scale = _Scale;

var _Selection = _interopRequireWildcard(require("./selection"));

exports.Selection = _Selection;

var _Style = _interopRequireWildcard(require("./style"));

exports.Style = _Style;

var _TextSize = _interopRequireWildcard(require("./textsize"));

exports.TextSize = _TextSize;

var _timer = _interopRequireDefault(require("./timer"));

var _timerContext = _interopRequireDefault(require("./timer-context"));

var _Transitions = _interopRequireWildcard(require("./transitions"));

exports.Transitions = _Transitions;

var _UserProps = _interopRequireWildcard(require("./user-props"));

exports.UserProps = _UserProps;

var _Wrapper = _interopRequireWildcard(require("./wrapper"));

exports.Wrapper = _Wrapper;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
