// victory-line/lib/v37/victory-line.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VictoryLine = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _curve = require("../curve");

var _victoryCore = require("victory-core");

var _v = require("victory-core/es/v37");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var VictoryLine = (0, _v.withNormalizedProps)({
  displayName: "VictoryLine",
  propTypes: {
    interpolation: _propTypes.default.oneOfType([_propTypes.default.oneOf(["basis", "bundle", "cardinal", "catmullRom", "linear", "monotoneX", "monotoneY", "natural", "step", "stepAfter", "stepBefore"]), _propTypes.default.func])
  },
  defaultProps: {
    width: 450,
    height: 300,
    padding: 50,
    data: [{
      x: 1,
      y: 1
    }, {
      x: 2,
      y: 2
    }, {
      x: 3,
      y: 3
    }, {
      x: 4,
      y: 4
    }],
    containerComponent: /*#__PURE__*/_react.default.createElement(_victoryCore.VictoryContainer, null),
    dataComponent: /*#__PURE__*/_react.default.createElement(_curve.Curve, null),
    interpolation: "linear",
    labelComponent: /*#__PURE__*/_react.default.createElement(_victoryCore.VictoryLabel, {
      renderInPortal: true
    }),
    groupComponent: /*#__PURE__*/_react.default.createElement(_victoryCore.VictoryClipContainer, null),
    samples: 50,
    sortKey: "x",
    sortOrder: "ascending",
    standalone: true,
    theme: _victoryCore.VictoryTheme.grayscale
  }
}, function (props) {
  return /*#__PURE__*/_react.default.createElement(_v.Clone, {
    element: props.groupComponent
  }, /*#__PURE__*/_react.default.createElement(_v.Clone, _extends({
    element: props.dataComponent
  }, props)));
});
exports.VictoryLine = VictoryLine;
