// victory-line/es/v37/victory-line.stories.js
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

/* eslint-disable react/no-multi-comp */

/* eslint-disable no-magic-numbers */
import * as React from "react";
import { VictoryLine } from "./victory-line";
import VictoryChart from "victory-chart/es/v37/victory-chart";
import { getData } from "../../../../stories/data";
export default {
  title: "v37/VictoryLine",
  parameters: {
    chromatic: {
      disableSnapshot: true
    }
  },
  component: VictoryLine
};
export var Demo = function (props) {
  return /*#__PURE__*/React.createElement(VictoryLine, _extends({}, props, {
    data: getData(8)
  }));
};
Demo.args = {
  lineRatio: 0.5
};
export var WithChart = function (props) {
  return /*#__PURE__*/React.createElement(VictoryChart, props, /*#__PURE__*/React.createElement(VictoryLine, null));
};
