// victory-bar/es/v37/victory-bar-v2.js
var _excluded = ["alignment", "barRatio", "barWidth", "cornerRadius", "data", "dataComponent", "groupComponent", "includeZero", "horizontal", "sortOrder"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/* eslint-disable react/no-multi-comp */
import * as React from "react";
import { useData, useDomain, useScale, useVictoryProviderSync, withContainer } from "victory-core/es/v37";
import Bar from "../bar";
import { getBarPosition } from "../helper-methods";

function VictoryBar(_ref) {
  var _ref$alignment = _ref.alignment,
      alignment = _ref$alignment === void 0 ? "middle" : _ref$alignment,
      barRatio = _ref.barRatio,
      barWidth = _ref.barWidth,
      cornerRadius = _ref.cornerRadius,
      _ref$data = _ref.data,
      data = _ref$data === void 0 ? [{
    x: 1,
    y: 1
  }, {
    x: 2,
    y: 2
  }, {
    x: 3,
    y: 3
  }, {
    x: 4,
    y: 4
  }] : _ref$data,
      _ref$dataComponent = _ref.dataComponent,
      dataComponent = _ref$dataComponent === void 0 ? /*#__PURE__*/React.createElement(Bar, null) : _ref$dataComponent,
      _ref$groupComponent = _ref.groupComponent,
      groupComponent = _ref$groupComponent === void 0 ? /*#__PURE__*/React.createElement("g", {
    role: "presentation"
  }) : _ref$groupComponent,
      _ref$includeZero = _ref.includeZero,
      includeZero = _ref$includeZero === void 0 ? true : _ref$includeZero,
      _ref$horizontal = _ref.horizontal,
      horizontal = _ref$horizontal === void 0 ? false : _ref$horizontal,
      _ref$sortOrder = _ref.sortOrder,
      sortOrder = _ref$sortOrder === void 0 ? "ascending" : _ref$sortOrder,
      props = _objectWithoutProperties(_ref, _excluded);

  // Typescript works best when we provide defaults in the destructured
  // component props. However, I am anticipating confusion about which props
  // we will need to pass back up to the provider vs. the props that are only
  // used in this component.
  useVictoryProviderSync(_objectSpread({
    data: data,
    includeZero: includeZero,
    sortOrder: sortOrder
  }, props));
  var scale = useScale();
  var domain = useDomain(); // TODO: Do we need to get this from context?

  var formattedData = useData();
  var barPositionProps = {
    domain: domain,
    scale: scale,
    horizontal: horizontal
  };
  var children = formattedData.map(function (datum, i) {
    var _getBarPosition = getBarPosition(barPositionProps, datum),
        x = _getBarPosition.x,
        y = _getBarPosition.y,
        y0 = _getBarPosition.y0;

    var dataProps = {
      index: i,
      key: "bar-".concat(i),
      alignment: alignment,
      barRatio: barRatio,
      barWidth: barWidth,
      cornerRadius: cornerRadius,
      scale: scale,
      data: data,
      x: x,
      y: y,
      y0: y0,
      datum: datum
    };
    return /*#__PURE__*/React.cloneElement(dataComponent, dataProps);
  });
  return /*#__PURE__*/React.cloneElement(groupComponent, {}, children);
} // @ts-expect-error we need to work out what props this withContainer wrapper accepts


export default withContainer(VictoryBar);
