// victory-bar/es/v37/victory-bar-v1.js
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/* eslint-disable react/no-multi-comp */
import * as React from "react";
import { VictoryContainer } from "victory-core";
import { useData, useDomain, useScale, VictoryProvider } from "victory-core/es/v37";
import Bar from "../bar";
import { getBarPosition } from "../helper-methods"; // This is a demo component that uses VictoryProvider to access calculated props and state.
// This component does not include events, animations, or styling.
// To test out, swap out the VictoryBar export in ./index.js and run `pnpm run storybook:server`.

var defaultProps = {
  data: [{
    x: 1,
    y: 1
  }, {
    x: 2,
    y: 2
  }, {
    x: 3,
    y: 3
  }, {
    x: 4,
    y: 4
  }],
  height: 300,
  includeZero: true,
  padding: 50,
  sortOrder: "ascending",
  width: 450
};
// TODO: This would be a shared helper that allows us to access context values in the base component
export function withContainer(WrappedComponent) {
  var initialProviderProps = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return function (props) {
    var providerProps = _objectSpread(_objectSpread({}, initialProviderProps), props);

    var _props$standalone = props.standalone,
        standalone = _props$standalone === void 0 ? true : _props$standalone,
        _props$containerCompo = props.containerComponent,
        containerComponent = _props$containerCompo === void 0 ? /*#__PURE__*/React.createElement(VictoryContainer, null) : _props$containerCompo;

    if (standalone) {
      return /*#__PURE__*/React.createElement(VictoryProvider, providerProps, /*#__PURE__*/React.cloneElement(containerComponent, providerProps, /*#__PURE__*/React.createElement(WrappedComponent, props)));
    }

    return /*#__PURE__*/React.createElement(VictoryProvider, providerProps, /*#__PURE__*/React.createElement(WrappedComponent, props));
  };
}

function VictoryBar(_ref) {
  var _ref$dataComponent = _ref.dataComponent,
      dataComponent = _ref$dataComponent === void 0 ? /*#__PURE__*/React.createElement(Bar, null) : _ref$dataComponent,
      _ref$groupComponent = _ref.groupComponent,
      groupComponent = _ref$groupComponent === void 0 ? /*#__PURE__*/React.createElement("g", {
    role: "presentation"
  }) : _ref$groupComponent,
      alignment = _ref.alignment,
      barRatio = _ref.barRatio,
      barWidth = _ref.barWidth,
      cornerRadius = _ref.cornerRadius,
      _ref$horizontal = _ref.horizontal,
      horizontal = _ref$horizontal === void 0 ? false : _ref$horizontal;
  var scale = useScale();
  var data = useData();
  var domain = useDomain();
  var children = data.map(function (datum, i) {
    var _getBarPosition = getBarPosition({
      domain: domain,
      scale: scale,
      horizontal: horizontal
    }, datum),
        x = _getBarPosition.x,
        y = _getBarPosition.y,
        x0 = _getBarPosition.x0,
        y0 = _getBarPosition.y0;

    var dataProps = {
      index: i,
      key: "bar-".concat(i),
      alignment: alignment,
      barRatio: barRatio,
      barWidth: barWidth,
      cornerRadius: cornerRadius,
      scale: scale,
      data: data,
      x: x,
      y: y,
      x0: x0,
      y0: y0,
      datum: datum
    };
    return /*#__PURE__*/React.cloneElement(dataComponent, dataProps);
  });
  return /*#__PURE__*/React.cloneElement(groupComponent, {}, children);
}

export default withContainer(VictoryBar, defaultProps);
